File: src/main/java/com/andredonadon/string/RunLengthEncoding.java
package com.andredonadon.string;

public class RunLengthEncoding {

    public static String encode(String input) {
        if (input == null || input.isEmpty()) {
            return "";
        }

        StringBuilder encoded = new StringBuilder();
        int count = 1;

        for (int i = 1; i < input.length(); i++) {
            if (input.charAt(i) == input.charAt(i - 1)) {
                count++;
            } else {
                encoded.append( input.charAt(i -1) ).append(count);
                count = 1;
            }
        }

        encoded.append(input.charAt(input.length() - 1)).append(count);

        return encoded.toString();
    }

    public static String decode(String encoded) {
        if (encoded == null || encoded.isEmpty()) {
            return "";
        }

        StringBuilder decoded = new StringBuilder();

        for (int i = 0; i < encoded.length(); i++) {
            char character = encoded.charAt(i);
            int count = 0;

            while ( (i+1 < encoded.length()) && (Character.isDigit(encoded.charAt(i+1)))) {
                count = count * 10 + (encoded.charAt(++i)) - '0';
            }

            decoded.append(String.valueOf(character).repeat(count));
        }

        return decoded.toString();
    }

    private static String reduced(String input) {
        if ((input == null) || input.isEmpty()) {
            return "";
        }

        StringBuilder sb = new StringBuilder();
        sb.append(input.charAt(0));

        for (int i = 1; i < input.length(); i++) {
            if (input.charAt(i) != input.charAt(i-1)) {
                sb.append(input.charAt(i));
            }
        }

        return sb.toString();
    }

    public static void main(String[] args) {
        String original = "aannndrrrrreeeee lluuuiiiizzzzzzz";
        String compressed = encode(original);
        String reduced = reduced(original);
        String decompressed = decode(compressed);

        System.out.println("Original:    " + original);
        System.out.println("Compactado:  " + compressed);
        System.out.println("Reduzido:  " + reduced);
        System.out.println("Descompactado: " + decompressed);
    }

}

File: kalgorithms/src/main/kotlin/com/andredonadon/kalgorithms/string/RunLengthEncoding.kt
package com.andredonadon.kalgorithms.string

fun encodeRLE(input: String) : String {
    if (input.isEmpty()) return ""

    val encoded = StringBuilder()
    var count = 1

    for (i in 1 until input.length) {
        if (input[i] == input[i-1]) {
            count++
        } else {
            encoded.append("${input[i-1]}$count")
            count = 1
        }
    }

    encoded.append("${input.last()}$count")

    return encoded.toString()
}

fun decodeRLE(encoded: String) : String {
    if (encoded.isEmpty()) return ""

    val regex = Regex("([a-zA-Z ])([0-9]+)")
    val decoded = StringBuilder()

    regex.findAll(encoded).forEach { match ->
        val char = match.groupValues[1]
        val count = match.groupValues[2].toInt()
        decoded.append(char.repeat(count))
    }

    return decoded.toString()
}

fun removeConsecutiveDuplicates(input: String) : String {
    if (input.isEmpty()) return ""

    val result = StringBuilder()
    result.append(input[0])

    for (i in 1 until input.length) {
        if (input[i] != input[i-1]) {
            result.append(input[i])
        }
    }

    return result.toString()
}

fun main() {
    val original = "aannndrrrrreeeee lluuuiiiizzzzzzz"
    val compressed = encodeRLE(original)
    val decompressed = decodeRLE(compressed)
    val reduced = removeConsecutiveDuplicates(original)


    println("Original:    $original")
    println("Compactado:  $compressed")
    println("Descompactado: $decompressed")
    println("Reduzido: $reduced")
    println("Reconstrução correta? ${original == decompressed}")
}